### PR TITLE
set zoom factor after page has finished loading

### LINKED ISSUE #770: Zooming function is functionally flawed
OS: elementary OS, Ubuntu 20.04-based
Kiwix-desktop version: 2.0.5, flatpak edition from flathub

My main issue is that, using the search function, the zoom level gets reset to 100% by Kiwix.
Therefore, even if I have set the default zoom at 130%, that will apply only to the homepage of the Wiki loaded on Kiwix, and the hyperlinks clicked from there. But it will *not* work for any page belonging to the same Wiki when accessed through the search function.

For all users who search wiki pages, rather than navigating them through hyperlinks only, this is a major obstacle to usage of the software.
Please refer to the video below for a demonstration of the behaviour

https://user-images.githubusercontent.com/16632292/150642804-668b4277-a4c7-47c6-915b-642a942a920a.mp4

### PR DESCRIPTION
**Pull request by juuz0**
Fixes #770

### CODE REVIEW

**Comment by juuz0** [comment]
I couldn't get a good reason on why that was happening (it didnt happen, if the search bar article was opened in a different tab) but this looks like the culprit, known bug :
https://bugreports.qt.io/browse/QTBUG-51992

**Review by kelson42 — CHANGES_REQUESTED** [review]
The search result page should have the zoom from the "default zoom" settings

**Comment by kelson42** [comment]
It seems we can not avoid this zoom changing problem because of https://bugreports.qt.io/browse/QTBUG-51851

**Review by kelson42 — APPROVED** [review]
Beside the rezooming bug (upstream) after clicking on a search result, it looks good. @mgautierfr Please make the code review.

**Comment by kelson42** [comment]
@juuz0 Would you be able please to make a comment in the code with a link to the upstream ticket to explain why you rezzoom after having loaded the content

**Comment by juuz0** [comment]
@kelson42 Added, rebased

**Comment by kelson42** [comment]
@veloman-yunkan Changing reviewe in an attemos to move forward with this PR :)

File: src/webview.cpp [hunk 1]
```diff
@@ -49,6 +49,25 @@ WebView::WebView(QWidget *parent)
...
+    connect(this, &QWebEngineView::loadFinished, this, [=] (bool ok) {
+        if (ok) {
+            auto url = this->url();
+            auto settingsManager = KiwixApp::instance()->getSettingsManager();
+            qreal zoomFactor;
+            QUrlQuery query(url);
+            if (query.hasQueryItem("pattern")) {
+               zoomFactor = settingsManager->getZoomFactor();
+            } else {
+                auto zimId = url.host().split('.')[0];
```
Comment (veloman-yunkan) [inline src/webview.cpp:65]: I feel suspicious about a ZIM id obtained in such a way. I see that it is a copy from `WebView::onUrlChanged()`. I think that we have a bug there. For example, the URL `http://library.kiwix.org/phet_hy_2021-08` will have `library` extracted from it as `zimId` whereas my impression was that we are interested in the `phet_hy_2021-08` part.

@kelson42 Will you please confirm that my understanding is correct?

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]

File: src/webview.cpp (hunk 1 above)
Comment (mgautierfr) [reply src/webview.cpp:65]: On kiwix-desktop, the url format is `zimid.zim/article/path`.
This is somehow a stable fact as the custom url handler is installed to respond to all `zim` tld (Other request are not catch, and so we can transfer to native browser for external url)
Extracting the zimId from the url could (should) be move to a dedicated function, but it is a correct way to obtain the zimId (In the customSchemeHandler, we simply remove the 4 last chars)

File: src/webview.cpp [hunk 2]
```diff
@@ -49,6 +49,25 @@ WebView::WebView(QWidget *parent)
...
+    /* This rezooms the page to its correct zoom(default/by ZIM ID) after loading is finished.
+    If the page is result of fulltext search, use the default zoom
+    Needed because of https://bugreports.qt.io/browse/QTBUG-51851 */
+    connect(this, &QWebEngineView::loadFinished, this, [=] (bool ok) {
+        if (ok) {
+            auto url = this->url();
+            auto settingsManager = KiwixApp::instance()->getSettingsManager();
+            qreal zoomFactor;
+            QUrlQuery query(url);
+            if (query.hasQueryItem("pattern")) {
```
[the remark this answers is not in the document]
Comment (juuz0) [reply src/webview.cpp:62]: Probably, my reasoning was in `suggestionlistworker.cpp`, it was always putting pattern query item

**Comment by kelson42** [comment]
@juuz0 Whats up with this PR?

File: src/webview.cpp [hunk 3]
```diff
@@ -40,6 +40,19 @@ void WebViewForwardMenu::showEvent(QShowEvent *)
...
 
+void WebView::applyCorrectZoomFactor() {
+    auto url = this->url();
+    auto settingsManager = KiwixApp::instance()->getSettingsManager();
+    qreal zoomFactor;
+    QUrlQuery query(url);
+    if (query.hasQueryItem("pattern")) {
+        zoomFactor = settingsManager->getZoomFactor();
+    } else {
+        auto zimId = url.host().split('.')[0];
```
Comment (veloman-yunkan) [inline src/webview.cpp:51]: Please follow @mgautierfr's hint in his [earlier comment](https://github.com/kiwix/kiwix-desktop/pull/774#discussion_r803707578)

> Extracting the zimId from the url could (should) be move to a dedicated function

File: src/webview.cpp [hunk 4]
```diff
@@ -40,6 +40,19 @@ void WebViewForwardMenu::showEvent(QShowEvent *)
     setGeometry(geo);
 }
 
+void WebView::applyCorrectZoomFactor() {
+    auto url = this->url();
+    auto settingsManager = KiwixApp::instance()->getSettingsManager();
+    qreal zoomFactor;
+    QUrlQuery query(url);
+    if (query.hasQueryItem("pattern")) {
```
Comment (veloman-yunkan) [inline src/webview.cpp:48]: The intent of this check will be more explicit (and the code more readable) if you encapsulate it into a helper function `WebView::isSearchResultsView()` or introduce an auxiliary variable in this function:

```
const bool isSearchResultsView = QUrlQuery(url).hasQueryItem("pattern");
if ( isSearchResultsView ) {
    ...
```

Besides, I still think that relying only on the presence of a parameter named  `pattern` in the URL query string is a weak criterion that can easily break if we happen to introduce another type of a URL with a similar parameter. Probably it makes sense to also check that the host component of the URL ends with ".search".

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]

File: src/webview.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (kelson42) [reply src/webview.cpp]: @juuz0 Should we close this discussion? What is the status?

[the remark this answers is not in the document]
Comment (juuz0) [reply src/webview.cpp]: @kelson42 Fixed following the discussion in https://github.com/kiwix/kiwix-desktop/pull/774#discussion_r814166559

**Review by veloman-yunkan — APPROVED** [review]

**Pull request merged**